library/src/hcc_detail/hipsolver.cpp

```diff
@@ -2902,6 +2958,664 @@ catch(...)
+                                             hip2rocblas_evect(jobz),
+                                             hip2rocblas_fill(uplo),
+                                             n,
+                                             nullptr,
+                                             lda,
+                                             nullptr,
+                                             nullptr,
+                                             nullptr);
+    rocblas_stop_device_memory_size_query((rocblas_handle)handle, &sz);
+
+    // space for E array
+    sz += sizeof(float) * n;
+
+    if(sz > INT_MAX)
+        return HIPSOLVER_STATUS_INTERNAL_ERROR;
+
+    *lwork = (int)sz;
+    return rocblas2hip_status(status);
+}
+catch(...)
+{
+    return exception2hip_status();
+}
+
+hipsolverStatus_t hipsolverZheevd_bufferSize(hipsolverHandle_t       handle,
+                                             hipsolverEigMode_t      jobz,
+                                             hipsolverFillMode_t     uplo,
+                                             int                     n,
+                                             hipsolverDoubleComplex* A,
+                                             int                     lda,
+                                             double*                 W,
+                                             int*                    lwork)
+try
+{
+    size_t sz;
+
+    rocblas_start_device_memory_size_query((rocblas_handle)handle);
+    rocblas_status status = rocsolver_zheevd((rocblas_handle)handle,
+                                             hip2rocblas_evect(jobz),
+                                             hip2rocblas_fill(uplo),
+                                             n,
+                                             nullptr,
+                                             lda,
+                                             nullptr,
+                                             nullptr,
+                                             nullptr);
+    rocblas_stop_device_memory_size_query((rocblas_handle)handle, &sz);
+
+    // space for E array
+    sz += sizeof(double) * n;
+
+    if(sz > INT_MAX)
+        return HIPSOLVER_STATUS_INTERNAL_ERROR;
+
+    *lwork = (int)sz;
+    return rocblas2hip_status(status);
+}
+catch(...)
+{
+    return exception2hip_status();
+}
+
+hipsolverStatus_t hipsolverSsyevd(hipsolverHandle_t   handle,
+                                  hipsolverEigMode_t  jobz,
+                                  hipsolverFillMode_t uplo,
+                                  int                 n,
+                                  float*              A,
+                                  int                 lda,
+                                  float*              W,
+                                  float*              work,
+                                  int                 lwork,
+                                  int*                devInfo)
+try
+{
+    float* E = work;
```

Comment:
how is this going to work if a user passes a nullptr because wanted to use rocsolver automatic workspace management?
You cannot pass E as nullptr to rocsolver_syevd

Reply:
Oh, you're right! That complicates things...

I guess we have a couple options.
1. Add E as an argument to the function. On the CUDA side it will do nothing.
2. If work is a null pointer, use `rocblas_device_malloc` to allocate internal space for E. I'm not sure how `rocblas_device_malloc` will respond to being used in both hipSOLVER and rocSOLVER.
3. Create a private API for syevd and sygvd in rocSOLVER that does not need E.

Reply:
I would try option 2 first. I believe it should work... at the beginning, before rocblas team added the `rocblas_device_malloc` wrapper, we were allocating workspace arrays one by one. 
If this does not work, I suggest option 1 then. 

(just to clarify, in rocSOLVER we cannot have E as part of the workspace because when info!=0, this array will contain information of the equivalent resulting system that could be useful in some applications. If E is part of the internally managed workspace we cannot guarantee it will persist after the call to the rocsolver function returns)

Reply:
Option 2 generally seems to work well, but I was able to trigger a failure case by allocating a large amount of memory with `rocblas_device_malloc` after E had already been allocated (the pointer to E is likely invalidated by rocBLAS reallocating the memory space). I'm not sure how big the problem has to be in order to trigger this failure case, since daily_lapack seemed to run fine.

I'd prefer to avoid using option 1. Perhaps option 2 can be salvaged by using `hipMalloc` instead of `rocblas_device_malloc`. I know the philosophy of hipBLAS does not allow calls to `hipMalloc`; do we want to follow the same philosophy in hipSOLVER?

Reply:
> I was able to trigger a failure case by allocating a large amount of memory with `rocblas_device_malloc` after E had already been allocated (the pointer to E is likely invalidated by rocBLAS reallocating the memory space)

That's unfortunate. I don't think there's any reason why that would have to be the case, but the rocblas allocator is probably too complex to change in time the next release cut-off.